File: system_simulations/network_model/java/discrete_event/Network.java
import java.util.ArrayList;

public class Network {
	private ArrayList<AtomicModel> components;
	private ArrayList<AtomicModel> inputs;
	private ArrayList<AtomicModel> outputs;
	private int componentCount;

	private Network(){}

	public static NetworkBuilder builder(){
		return new NetworkBuilder();
	}

	public void addInputToOutput(AtomicModel I, AtomicModel O){
		I.addOutput(O);
		O.addInput(I);
	}

	public ArrayList<AtomicModel> getComponents(){
		return components;
	}

	public ArrayList<AtomicModel> getInputs(){
		return inputs;
	}

	public ArrayList<AtomicModel> getOutputs(){
		return outputs;
	}

	public static class NetworkBuilder{
		private ArrayList<AtomicModel> components = new ArrayList<AtomicModel>();
		private ArrayList<AtomicModel> inputs = new ArrayList<AtomicModel>();
		private ArrayList<AtomicModel> outputs = new ArrayList<AtomicModel>();
		private int componentCount = 0;

		public NetworkBuilder(){}

		public NetworkBuilder addComponent(AtomicModel component){
			this.components.add(component);
			return this;
		}

		public NetworkBuilder addInput(AtomicModel I){
			this.inputs.add(I);
			return this;
		}

		public NetworkBuilder addOutput(AtomicModel O){
			this.outputs.add(O);
			return this;
		}

		public Network build(){
			Network network = new Network();
			network.components = this.components;
			network.componentCount = this.components.size();
			network.inputs = this.inputs;
			network.outputs = this.outputs;
			return network;
		}
	}
}

File: system_simulations/network_model/cpp/discrete_event/Network.cpp
#ifndef NETWORK
#define NETWORK
#include <vector>
#include "AtomicModel.cpp"
using namespace std;

template<class T>
class Network {
	public:
		class NetworkBuilder {
			public:
				NetworkBuilder(){
				}

				NetworkBuilder& add_component(T* component){
					this->components.push_back(component);
					return *this;
				}

				NetworkBuilder& add_input(T* input){
					this->inputs.push_back(input);
					return *this;
				}

				NetworkBuilder& add_output(T* output){
					this->outputs.push_back(output);
					return *this;
				}

				Network* build(){
					return new Network(this);
				}

			private:
				vector<T*> components;
				vector<T*> inputs;
				vector<T*> outputs;
			friend Network;
		};

		Network(NetworkBuilder* builder){
			this->components = builder->components;
			this->inputs = builder->inputs;
			this->outputs = builder->outputs;
		}

		vector<T*> get_inputs(){
			return inputs;
		}

		vector<T*> get_outputs(){
			return outputs;
		}

		vector<T*> get_components(){
			return components;
		}

		string get_name(){
			return "Network";
		}
		
	private:
		vector<T*> components;
		vector<T*> inputs;
		vector<T*> outputs;
};

#endif